### PR TITLE
EjectAllOfType also clears WorkTracker (same as Eject does)

### PR DESCRIPTION
**Pull request by mronnblom-cint**
Fixes https://github.com/JasperFx/marten/issues/2005

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/JasperFx/marten?pullRequest=2010) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/JasperFx/marten?pullRequest=2010) before we can accept your contribution.<br/><br/>:x: mronnblom-cint [sign now](https://cla.dotnetfoundation.org/JasperFx/marten?pullRequest=2010)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/JasperFx/marten?pullRequest=2010) it.</sub>

**Review by mysticmind — APPROVED** [review]
Looks good to me 👍

**Review by oskardudycz — APPROVED** [review]

File: src/Marten/Internal/UnitOfWork.cs [hunk 1]
```diff
@@ -219,6 +219,17 @@ public void Eject<T>(T document)
             }
         }
 
+        public void EjectAllOfType(Type type)
+        {
+            var operations = operationsFor(type);
+            var matching = operations.OfType<IDocumentStorageOperation>().ToArray();
+
+            foreach (var operation in matching)
```
Comment (jeremydmiller) [inline src/Marten/Internal/UnitOfWork.cs:227]: You have to check whether the `IDocumentStorageOperation` involves the `type` here.

File: src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+using Marten.Testing.Harness;
+using Shouldly;
+using Xunit;
+
+namespace Marten.Testing.CoreFunctionality
+{
+    public class ejecting_a_document_type : IntegrationContext
+    {
+        [Fact]
+        public void eject_a_document_type_clears_it_from_the_identity_map_regular()
```
Comment (jeremydmiller) [inline src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: You probably want tests for Insert() and Update() as well as Store(). And you need to add operations from some sort of document besides `Target` to ensure that only the expected document type is impacted.

**Review by jeremydmiller — CHANGES_REQUESTED** [review]
This just needs the document type filtering and a test proving that.

File: src/Marten/Internal/UnitOfWork.cs (hunk 1 above)
Comment (mysticmind) [reply src/Marten/Internal/UnitOfWork.cs:227]: @jeremydmiller  `var operations = operationsFor(type);`, this line does that check, anything else needed?

Comment (jeremydmiller) [reply src/Marten/Internal/UnitOfWork.cs:227]: Ah, sorry, I missed that. Still need a test to prove that it's filtering correctly though.

Comment (mronnblom-cint) [reply src/Marten/Internal/UnitOfWork.cs:227]: Good point! Hopefully I have time later this week to add more tests!

File: src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs (hunk 2 above)
Comment (mronnblom-cint) [reply src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: I'm in the process of adding tests for `Update()`, but I don't understand why this test fails (even with `EjectAllOfType` commented out). I have never used the `DirtyTrackedSession` myself, so maybe I've misunderstood it, but shouldn't it track pending changes without needing to call `Update`?

```
[Fact]
public void eject_a_document_type_clears_updates_from_the_unit_of_work_dirty()
{
    var target1 = new Target { Number = 1 };
    var target2 = new Target { Number = 2 };
    var user1 = new User { Age = 10 };
    var user2 = new User { Age = 20 };

    using (var session = theStore.OpenSession())
    {
        session.Store(target1, target2);
        session.Store(user1, user2);

        session.SaveChanges();
    }

    using (var session = theStore.DirtyTrackedSession())
    {
        target1.Number = 3;
        target2.Number = 4;
        user1.Age = 30;
        user2.Age = 40;

        //session.EjectAllOfType(typeof(Target));

        session.SaveChanges();
    }

    using (var session = theStore.QuerySession())
    {
        session.Load<Target>(target1.Id).ShouldNotBeNull().Number.ShouldBe(1);
        session.Load<Target>(target2.Id).ShouldNotBeNull().Number.ShouldBe(2);
        session.Load<User>(user1.Id).ShouldNotBeNull().Age.ShouldBe(30);
        session.Load<User>(user2.Id).ShouldNotBeNull().Age.ShouldBe(40);
    }
}
```

Comment (jeremydmiller) [reply src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: Just what you said, the dirty tracking checks all documents that were loaded through the session at the time SaveChanges() was called. Very inefficient, only useful as a convenience when perf doesn't matter much.

Comment (mronnblom-cint) [reply src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: Ok. So why doesn't the test above work? When you do e.g. `target1.Number = 3` inside the `DirtyTrackedSession` and then `SaveChanges`, why is that change not committed?

Comment (mronnblom-cint) [reply src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: I just realized something myself.. Do you need to load the objects from the actual `DirtyTrackedSession` for it to know about them? That would probably make sense.

Comment (mronnblom-cint) [reply src/Marten.Testing/CoreFunctionality/ejecting_a_document_type.cs:11]: Yep, that was it. See latest commit.

**Comment by jeremydmiller** [comment]
I'm adding a little bit more to the tests locally, then pulling this in today.

**Pull request merged**